src/python-bindings/htcondor2/remote_param.cpp

```diff
@@ -0,0 +1,242 @@
+        return false;
+    }
+
+    // Not sure we actually need to make this copy, but version 1 did.
+    ClassAd copy;
+    copy.CopyFrom(location);
+
+    Daemon d( &copy, DT_GENERIC, NULL );
+
+    CondorError errorStack;
+    bool result = sock.connect( d.addr(), 0, false, & errorStack );
+    if(! result) {
+        // This was HTCondorValueError in version 1.
+        PyErr_SetString( PyExc_IOError, "Failed to connect to daemon." );
+        return false;
+    }
+
+    result = d.startCommand( cmd, & sock, 0, & errorStack );
+    if(! result) {
+        // This was HTCondorIOError in version 1.
+        PyErr_SetString( PyExc_IOError, "Failed to start command." );
+        return false;
+    }
+
+    return true;
+}
+
+
+static PyObject *
+_remote_param_keys( PyObject *, PyObject * args ) {
+    // _param_keys(self.location._handle)
+    PyObject_Handle * handle = NULL;
+
+    if(! PyArg_ParseTuple( args, "O", (PyObject **)& handle )) {
+        // PyArg_ParseTuple() has already set an exception for us.
+        return NULL;
+    }
+
+    auto * location = (ClassAd * )handle->t;
+
+    ReliSock sock;
+    bool result = start_config_command( DC_CONFIG_VAL, sock, * location );
+    if(! result) {
+        // start_config_command() has already set an exception for us.
+        return NULL;
+    }
+
+    sock.encode();
+
+    std::string payload = "?names";
+    if(! sock.put(payload)) {
+        // This was HTCondorIOError in version 1.
+        PyErr_SetString( PyExc_IOError, "Failed to send request for parameter names." );
+        return NULL;
+    }
+    if(! sock.end_of_message()) {
+        // This was HTCondorIOError in version 1.
+        PyErr_SetString( PyExc_IOError, "Failed to send EOM for parameter names." );
+        return NULL;
+    }
+
+    sock.decode();
+
+    std::string reply;
+    if(! sock.code(reply)) {
+        // This was HTCondorIOError in version 1.
+        PyErr_SetString( PyExc_IOError, "Failed to receive rely for parameter names." );
+        return NULL;
+    }
+
+    if( reply == "Not defined" ) {
+        if(! sock.end_of_message()) {
+            // This was HTCondorIOError in version 1.
+            PyErr_SetString( PyExc_IOError, "Failed to receive EOM from remote daemon (unsupported version)." );
+            return NULL;
+        }
+
+        // Strictly speaking, this could be a reply from a daemon earlier
+        // than version 8.1.2, but let's not add another round-trip to
+        // half-ass a check by looking for a known knob name.
+
+        // This was HTCondorReplyError in version 1.
+        PyErr_SetString( PyExc_IOError, "Not authorized to query remote daemon." );
+        return NULL;
+    }
+    if( reply[0] == '!' ) {
+        sock.end_of_message();
+
+        // This was HTCondorIOError in version 1.
+        PyErr_SetString( PyExc_IOError, "Remote daemon failed to get parameter names." );
+        return NULL;
+    }
+
+    StringList keys( NULL, "," );
```

Comment:
Considering ``StringList`` is on the way out the door could we change this to be a vector of strings that we join together.